In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import scipy.stats as stats
from scipy.stats import chi2_contingency
from scipy.stats import kruskal
import plotly.express as px

warnings.filterwarnings('ignore')

pd.set_option('display.max_columns',500)

def switch_plot_mode(mode):
    if mode == 'dark':
        plt.style.use('dark_background')
        plt.rcParams['figure.figsize'] = (12,6)
        plt.rcParams['font.size'] = 12
        plt.rcParams['text.color'] = 'white'
        plt.rcParams['axes.labelcolor'] = 'white'
        plt.rcParams['xtick.color'] = 'white'
        plt.rcParams['ytick.color'] = 'white'
    elif mode == 'light':
        plt.style.use('default')
        plt.rcParams['figure.figsize'] = (12,6)
        plt.rcParams['font.size'] = 12
        plt.rcParams['text.color'] = 'black'
        plt.rcParams['axes.labelcolor'] = 'black'
        plt.rcParams['xtick.color'] = 'black'
        plt.rcParams['ytick.color'] = 'black'

switch_plot_mode('light')
        


In [2]:
data = pd.read_csv("superstore_dataset.csv")
data

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales Price,Quantity,Discount,Profit
0,1,CA-2016-152156,08/11/16,11/11/16,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2.0,0.00,41.9136
1,2,CA-2016-152156,08/11/16,11/11/16,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3.0,0.00,219.5820
2,3,CA-2016-138688,12/06/16,16/06/16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2.0,0.00,6.8714
3,4,US-2015-108966,11/10/15,18/10/15,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5.0,0.45,-383.0310
4,5,US-2015-108966,11/10/15,18/10/15,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2.0,0.20,2.5164
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10009,1048,CA-2016-169103,08/03/16,13/03/16,Standard Class,ZC-21910,Zuschuss Carroll,Consumer,United States,Miami,Florida,33142,South,FUR-FU-10004006,Furniture,Furnishings,"Deflect-o DuraMat Lighweight, Studded, Beveled...",102.3600,3.0,0.20,-3.8385
10010,9613,CA-2016-110086,18/09/16,22/09/16,Standard Class,BD-11320,Bill Donatelli,Consumer,United States,Woodland,California,95695,West,TEC-PH-10001299,Technology,Phones,Polycom CX300 Desktop Phone USB VoIP phone,239.9840,2.0,0.20,23.9984
10011,9614,CA-2014-128524,11/11/14,2014-11-13,First Class,MZ-17515,Mary Zewe,Corporate,United States,Seattle,Washington,98115,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,22.9800,1.0,0.00,6.8940
10012,9615,CA-2014-128524,11/11/14,2014-11-13,First Class,MZ-17515,Mary Zewe,Corporate,United States,Seattle,Washington,98115,West,TEC-AC-10002718,Technology,Accessories,Belkin Standard 104 key USB Keyboard,102.1300,7.0,0.00,15.3195


In [3]:
# Exploratory Data Analysis (EDA) step - Data Cleaning


In [4]:
data.head(5)


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales Price,Quantity,Discount,Profit
0,1,CA-2016-152156,08/11/16,11/11/16,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2.0,0.00,41.9136
1,2,CA-2016-152156,08/11/16,11/11/16,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3.0,0.00,219.5820
2,3,CA-2016-138688,12/06/16,16/06/16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2.0,0.00,6.8714
3,4,US-2015-108966,11/10/15,18/10/15,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5.0,0.45,-383.0310
4,5,US-2015-108966,11/10/15,18/10/15,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2.0,0.20,2.5164


In [5]:
data.head()


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales Price,Quantity,Discount,Profit
0,1,CA-2016-152156,08/11/16,11/11/16,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2.0,0.00,41.9136
1,2,CA-2016-152156,08/11/16,11/11/16,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3.0,0.00,219.5820
2,3,CA-2016-138688,12/06/16,16/06/16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2.0,0.00,6.8714
3,4,US-2015-108966,11/10/15,18/10/15,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5.0,0.45,-383.0310
4,5,US-2015-108966,11/10/15,18/10/15,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2.0,0.20,2.5164


In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10014 entries, 0 to 10013
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         10014 non-null  int64  
 1   Order ID       10014 non-null  object 
 2   Order Date     10014 non-null  object 
 3   Ship Date      10014 non-null  object 
 4   Ship Mode      9916 non-null   object 
 5   Customer ID    10014 non-null  object 
 6   Customer Name  10014 non-null  object 
 7   Segment        10014 non-null  object 
 8   Country        10014 non-null  object 
 9   City           10014 non-null  object 
 10  State          10014 non-null  object 
 11  Postal Code    10014 non-null  int64  
 12  Region         10014 non-null  object 
 13  Product ID     10014 non-null  object 
 14  Category       10014 non-null  object 
 15  Sub-Category   10014 non-null  object 
 16  Product Name   10014 non-null  object 
 17  Sales Price    10014 non-null  float64
 18  Quanti

In [7]:
data.isnull().sum()

Row ID            0
Order ID          0
Order Date        0
Ship Date         0
Ship Mode        98
Customer ID       0
Customer Name     0
Segment           0
Country           0
City              0
State             0
Postal Code       0
Region            0
Product ID        0
Category          0
Sub-Category      0
Product Name      0
Sales Price       0
Quantity         18
Discount          0
Profit            0
dtype: int64

In [8]:
data.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales Price', 'Quantity', 'Discount', 'Profit'],
      dtype='object')

In [9]:
data.shape

(10014, 21)

In [10]:
data.dtypes

Row ID             int64
Order ID          object
Order Date        object
Ship Date         object
Ship Mode         object
Customer ID       object
Customer Name     object
Segment           object
Country           object
City              object
State             object
Postal Code        int64
Region            object
Product ID        object
Category          object
Sub-Category      object
Product Name      object
Sales Price      float64
Quantity         float64
Discount         float64
Profit           float64
dtype: object

In [11]:
data.describe()

,Row ID,Postal Code,Sales Price,Quantity,Discount,Profit
count,10014.000000,10014.000000,10014.000000,9996.000000,10014.000000,10014.000000
mean,4997.523367,55205.131616,230.038424,3.792517,0.156240,28.708317
std,2888.981754,32066.711027,623.057359,2.231021,0.206392,234.154272
min,1.000000,1040.000000,-31.500000,1.000000,0.000000,-6599.978000
25%,2494.250000,23223.000000,17.280000,2.000000,0.000000,1.731375
50%,4997.500000,57103.000000,54.780000,3.000000,0.200000,8.666500
75%,7500.750000,90008.000000,209.970000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


In [12]:
data.nunique()


Row ID           9994
Order ID         5009
Order Date       1240
Ship Date        1334
Ship Mode           4
Customer ID       793
Customer Name     793
Segment             3
Country             1
City              531
State              54
Postal Code       631
Region              4
Product ID       1862
Category            3
Sub-Category       17
Product Name     1850
Sales Price      5829
Quantity           14
Discount           12
Profit           7287
dtype: int64

Handling Duplicates

In [13]:
duplicate_rows_mask = data.duplicated()
dups = data.loc[duplicate_rows_mask]

print(f"Total number of rows: {data.shape[0]}")
print(f"Number of duplicate rows found: {len(dups)}")

Total number of rows: 10014
Number of duplicate rows found: 17


In [14]:
duplicate_rowid_mask = data.duplicated(["Row ID"])
print(f"Number of duplicate (entire) rows are - {len(data.loc[duplicate_rows_mask])}")
print(f"Number of rows with duplicate Row ID attribute are - {len(data.loc[duplicate_rowid_mask])}")

Number of duplicate (entire) rows are - 17
Number of rows with duplicate Row ID attribute are - 20


In [15]:
data.loc[duplicate_rowid_mask & -duplicate_rows_mask]

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales Price,Quantity,Discount,Profit
9998,20,CA-2014-143336,27/08/14,2014-9-1,Second Class,ZD-21925,Zuschuss Donatelli,Consumer,United States,San Francisco,California,94109,West,TEC-PH-10001949,Technology,Phones,Cisco SPA 501G IP Phone,213.480,3.0,0.2,16.0110
10000,8852,CA-2015-140718,02/07/15,04/07/15,First Class,FA-14230,Frank Atkinson,Corporate,United States,Raleigh,North Carolina,27604,South,OFF-PA-10002250,Office Supplies,Paper,Things To Do Today Pad,-9.392,2.0,0.2,3.2872
10005,1044,CA-2017-115651,09/07/17,12/07/17,First Class,NS-18640,Noel Staavos,Corporate,United States,Chicago,Illinois,60610,Central,OFF-AR-10001130,Office Supplies,Art,"Quartet Alpha White Chalk, 12/Pack",-8.840,5.0,0.2,2.9835


In [16]:
duplicate_rows_mask2 = data.duplicated(keep = "last")
duplicate_rowid_mask2 = data.duplicated(["Row ID"],keep = "last")
data.loc[duplicate_rowid_mask2 & -duplicate_rows_mask2]

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales Price,Quantity,Discount,Profit
19,20,CA-2014-143336,27/08/14,2014-9-1,Second Class,ZD-21925,Zuschuss Donatelli,Consumer,United States,San Francisco,California,94109,West,TEC-PH-10001949,Technology,Phones,Cisco SPA 501G IP Phone,213.480,NaN,0.2,16.0110
1043,1044,CA-2017-115651,09/07/17,12/07/17,First Class,NS-18640,Noel Staavos,Corporate,United States,Chicago,Illinois,60610,Central,OFF-AR-10001130,Office Supplies,Art,"Quartet Alpha White Chalk, 12/Pack",8.840,5.0,0.2,2.9835
8851,8852,CA-2015-140718,02/07/15,04/07/15,First Class,FA-14230,Frank Atkinson,Corporate,United States,Raleigh,North Carolina,27604,South,OFF-PA-10002250,Office Supplies,Paper,Things To Do Today Pad,9.392,2.0,0.2,3.2872


In [17]:
neg_sales_prices = data.loc[data['Sales Price']<0]
neg_sales_prices

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales Price,Quantity,Discount,Profit
9794,9795,CA-2014-127166,21/05/14,2014-5-23,Second Class,KH-16360,Katherine Hughes,Consumer,United States,Houston,TX,77070,Central,OFF-BI-10000977,Office Supplies,Binders,Ibico Plastic Spiral Binding Combs,-18.240,3.0,0.8,-31.0080
9984,9985,CA-2015-100251,17/05/15,23/05/15,Standard Class,DV-13465,Dianna Vittorini,Consumer,United States,Long Beach,NY,11561,East,OFF-LA-10003766,Office Supplies,Labels,Self-Adhesive Removable Labels,-31.500,10.0,0.0,15.1200
10000,8852,CA-2015-140718,02/07/15,04/07/15,First Class,FA-14230,Frank Atkinson,Corporate,United States,Raleigh,North Carolina,27604,South,OFF-PA-10002250,Office Supplies,Paper,Things To Do Today Pad,-9.392,2.0,0.2,3.2872
10005,1044,CA-2017-115651,09/07/17,12/07/17,First Class,NS-18640,Noel Staavos,Corporate,United States,Chicago,Illinois,60610,Central,OFF-AR-10001130,Office Supplies,Art,"Quartet Alpha White Chalk, 12/Pack",-8.840,5.0,0.2,2.9835


In [18]:
data['Sales Price'] = np.abs(data['Sales Price'])
neg_sales_prices = data.loc[data['Sales Price']<0]
neg_sales_prices

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales Price,Quantity,Discount,Profit


In [19]:
labels_to_drop = list(data.loc[duplicate_rowid_mask2].index)
data.drop(labels_to_drop, axis="index", inplace = True)
data.shape

(9994, 21)

> Dropping the Row Id column as its need is sufficed
> The dataframe index maintains a unique row number for each row same as row_id column

In [20]:
data.drop(["Row ID"], axis = 'columns', inplace = True)

In [21]:
# Do duplicate rows still exist?
duplicate_rows_mask = data.duplicated()
data[duplicate_rows_mask]

,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales Price,Quantity,Discount,Profit
3406,US-2014-150119,23/04/14,2014-4-27,Standard Class,LB-16795,Laurel Beltran,Home Office,United States,Columbus,Ohio,43229,East,FUR-CH-10002965,Furniture,Chairs,Global Leather Highback Executive Chair with P...,281.372,2.0,0.3,-12.0588


In [22]:
data.drop(data[duplicate_rows_mask].index, axis="index", inplace = True)
data.shape

(9993, 20)

In [23]:
data.reset_index(drop=True, inplace=True)

Handling Duplicates : Summary
findings:
 1. A total of 21 duplicate rows were identified and removed from the dataset.
 2. The total number of rows in the DataFrame was reduced from 10,014 to 9,993.
 3. Since the index would certain missing values, we reset the index again.

In [24]:
# Data Handling

In [25]:
data['Order Date'].unique()[0:100]

array(['08/11/16', '12/06/16', '11/10/15', '09/06/14', '15/04/17',
       '05/12/16', '22/11/15', '11/11/14', '09/12/16', '16/07/17',
       '25/09/15', '16/01/16', '17/09/15', '19/10/17', '08/12/16',
       '27/12/15', '10/09/17', '17/07/16', '19/09/17', '11/03/16',
       '20/10/14', '20/06/16', '18/04/15', '11/12/16', '17/06/16',
       '24/11/17', '30/04/15', '05/12/14', '04/06/16', '18/09/16',
       '14/09/17', '26/04/15', '09/12/17', '26/11/14', '12/10/14',
       '03/09/15', '13/11/17', '28/05/17', '26/10/17', '05/04/16',
       '17/09/16', '31/01/15', '06/11/17', '09/11/17', '17/06/17',
       '06/09/16', '29/08/16', '01/12/16', '13/11/15', '23/11/17',
       '15/10/15', '25/12/17', '03/11/16', '25/08/14', '02/03/15',
       '05/04/15', '26/12/14', '20/09/14', '05/11/17', '06/11/16',
       '02/02/17', '13/10/16', '05/09/16', '18/09/17', '22/12/17',
       '07/09/15', '22/10/14', '13/03/16', '31/05/15', '28/05/15',
       '01/03/14', '20/11/16', '11/05/16', '28/12/15', '16/11/

In [30]:
data['Order Date'] = pd.to_datetime(data['Order Date'], format = '%d/%m/%y')
data.dtypes

Order ID                 object
Order Date       datetime64[ns]
Ship Date                object
Ship Mode                object
Customer ID              object
Customer Name            object
Segment                  object
Country                  object
City                     object
State                    object
Postal Code               int64
Region                   object
Product ID               object
Category                 object
Sub-Category             object
Product Name             object
Sales Price             float64
Quantity                float64
Discount                float64
Profit                  float64
dtype: object

Therefore, all 9,993 Order Date got changed to Date Time

In [32]:
data['Ship Date'].unique()[0:100]

array(['11/11/16', '16/06/16', '18/10/15', '2014-6-14', '20/04/17',
       '10/12/16', '26/11/15', '2014-11-18', '13/12/16', '18/07/17',
       '30/09/15', '20/01/16', '21/09/15', '23/10/17', '31/12/15',
       '15/09/17', '22/07/16', '23/09/17', '13/03/16', '2014-10-25',
       '25/06/16', '22/04/15', '17/12/16', '18/06/16', '30/11/15',
       '05/05/15', '2014-12-10', '06/06/16', '23/09/16', '17/09/17',
       '02/05/15', '11/12/17', '2014-12-1', '15/06/16', '2014-10-16',
       '08/09/15', '16/11/17', '30/05/17', '02/11/17', '10/04/16',
       '22/09/16', '05/02/15', '12/11/17', '11/11/17', '20/06/17',
       '11/09/16', '02/09/16', '04/12/16', '17/11/15', '28/11/17',
       '20/10/15', '30/12/17', '10/11/16', '2014-8-27', '06/03/15',
       '10/04/15', '2014-12-28', '2014-9-25', '05/02/17', '19/10/16',
       '07/09/16', '27/12/17', '12/09/15', '2014-10-28', '09/12/16',
       '16/03/16', '02/06/15', '03/06/15', '2014-3-6', '24/11/16',
       '12/05/16', '20/11/16', '2014-9-12', '2

In [35]:
# Handling ship dates of type - dd/mm/yy
dt_type1_mask = data['Ship Date'].str.contains(r'\d{2}[/]\d{2}[/]\d{2}', regex = True, na=False)
data.loc[dt_type1_mask, 'Ship Date']

0       11/11/16
1       11/11/16
2       16/06/16
3       18/10/15
4       18/10/15
          ...   
9985    12/07/17
9986    16/10/17
9987    13/03/16
9988    13/03/16
9989    22/09/16
Name: Ship Date, Length: 8054, dtype: object

In [37]:
data.loc[dt_type1_mask, 'Ship Date'] = pd.to_datetime(data.loc[dt_type1_mask, 'Ship Date'], format = '%d/%m/%y' )
data.loc[dt_type1_mask, 'Ship Date']

0       2016-11-11 00:00:00
1       2016-11-11 00:00:00
2       2016-06-16 00:00:00
3       2015-10-18 00:00:00
4       2015-10-18 00:00:00
               ...         
9985    2017-07-12 00:00:00
9986    2017-10-16 00:00:00
9987    2016-03-13 00:00:00
9988    2016-03-13 00:00:00
9989    2016-09-22 00:00:00
Name: Ship Date, Length: 8054, dtype: object

In [39]:
data = data.astype({'Ship Date': 'datetime64[ns]'})
data.dtypes

Order ID                 object
Order Date       datetime64[ns]
Ship Date        datetime64[ns]
Ship Mode                object
Customer ID              object
Customer Name            object
Segment                  object
Country                  object
City                     object
State                    object
Postal Code               int64
Region                   object
Product ID               object
Category                 object
Sub-Category             object
Product Name             object
Sales Price             float64
Quantity                float64
Discount                float64
Profit                  float64
dtype: object

Check Whether or not the years in Order Date and Order Id match

In [43]:
# Order Id : xx-YYY-xxxxx
order_id_yrs = data['Order ID'].str.extract(r'-(\d{2,4})-', expand = False).astype('int')
order_date_yrs = data['Order Date'].dt.year

inconsistent_order_dates_mask = (order_id_yrs != order_date_yrs)

inconsistent_order_dates = data[inconsistent_order_dates_mask]
print(f'# records with inconsistent order dates are - {inconsistent_order_dates.shape[0]}')

# records with inconsistent order dates are - 40


In [44]:
inconsistent_order_dates

,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales Price,Quantity,Discount,Profit
58,CA-2015-135545,2017-11-24,2015-11-30,Standard Class,KM-16720,Kunst Miller,Consumer,United States,Los Angeles,California,90004,West,TEC-AC-10004633,Technology,Accessories,Verbatim 25 GB 6x Blu-ray Single Layer Recorda...,13.980,2.0,0.0,6.1512
59,CA-2015-135545,2017-11-24,2015-11-30,Standard Class,KM-16720,Kunst Miller,Consumer,United States,Los Angeles,California,90004,West,OFF-BI-10001078,Office Supplies,Binders,"Acco PRESSTEX Data Binder with Storage Hooks, ...",25.824,6.0,0.2,9.3612
60,CA-2015-135545,2017-11-24,2015-11-30,Standard Class,KM-16720,Kunst Miller,Consumer,United States,Los Angeles,California,90004,West,OFF-PA-10003892,Office Supplies,Paper,Xerox 1943,146.730,3.0,0.0,68.9631
61,CA-2015-135545,2017-11-24,2015-11-30,Standard Class,KM-16720,Kunst Miller,Consumer,United States,Los Angeles,California,90004,West,FUR-FU-10000397,Furniture,Furnishings,Luxo Economy Swing Arm Lamp,79.760,4.0,0.0,22.3328
939,CA-2015-120880,2013-05-29,2015-06-03,Standard Class,JL-15850,John Lucas,Consumer,United States,Seattle,Washington,98105,West,OFF-PA-10004101,Office Supplies,Paper,Xerox 1894,32.400,5.0,0.0,15.5520
940,CA-2015-120880,2013-05-29,2015-06-03,Standard Class,JL-15850,John Lucas,Consumer,United States,Seattle,Washington,98105,West,OFF-ST-10001496,Office Supplies,Storage,Standard Rollaway File with Lock,540.570,3.0,0.0,140.5482
941,CA-2015-120880,2013-05-29,2015-06-03,Standard Class,JL-15850,John Lucas,Consumer,United States,Seattle,Washington,98105,West,OFF-BI-10002931,Office Supplies,Binders,"Avery Trapezoid Extra Heavy Duty 4"" Binders",167.760,5.0,0.2,62.9100
1142,CA-2017-160395,2016-12-31,2017-01-06,Standard Class,KL-16555,Kelly Lampkin,Corporate,United States,Reno,Nevada,89502,West,OFF-AR-10003759,Office Supplies,Art,"Crayola Anti Dust Chalk, 12/Pack",3.640,2.0,0.0,1.6380
1143,CA-2017-160395,2016-12-31,2017-01-06,Standard Class,KL-16555,Kelly Lampkin,Corporate,United States,Reno,Nevada,89502,West,OFF-BI-10003984,Office Supplies,Binders,Lock-Up Easel 'Spel-Binder',159.768,7.0,0.2,53.9217
2742,CA-2015-155306,1999-04-17,2015-04-21,Standard Class,GA-14515,George Ashbrook,Consumer,United States,San Francisco,California,94122,West,OFF-AR-10003251,Office Supplies,Art,Prang Drawing Pencil Set,5.560,2.0,0.0,2.2240


Replacing the years in Order Date with years in Order ID

In [47]:
data['year_diff'] = order_id_yrs - order_date_yrs
data['Order Date'] = data.apply(lambda x: x['Order Date'] + pd.DateOffset(years = x['year_diff']), axis = 'columns')

In [49]:
data.loc[inconsistent_order_dates_mask, ['Order ID', 'Order Date']]

,Order ID,Order Date
58,CA-2015-135545,2015-11-24
59,CA-2015-135545,2015-11-24
60,CA-2015-135545,2015-11-24
61,CA-2015-135545,2015-11-24
939,CA-2015-120880,2015-05-29
940,CA-2015-120880,2015-05-29
941,CA-2015-120880,2015-05-29
1142,CA-2017-160395,2017-12-31
1143,CA-2017-160395,2017-12-31
2742,CA-2015-155306,2015-04-17


In [51]:
data.drop(columns=['year_diff'], inplace = True)

In [53]:
print(data['Category'].unique())
print(data['Sub-Category'].unique())

['Furniture' 'Office Supplies' 'Technology']
['Bookcases' 'Chairs' 'Labels' 'Tables' 'Storage' 'Furnishings' 'Art'
 'Phones' 'Binders' 'Appliances' 'Paper' 'Accessories' 'Envelopes'
 'Fasteners' 'Supplies' 'Machines' 'Copiers']


In [57]:
# Regex
data_cat_codes = data["Product ID"].str.extract(r'^([A-Z]{3})-([A-Z]{2})-\d+$')
data_cat_codes.rename(columns={0:'cat_code', 1:'sub_cat_code'}, inplace=True)
data_cat_codes

,cat_code,sub_cat_code
0,FUR,BO
1,FUR,CH
2,OFF,LA
3,FUR,TA
4,OFF,ST
...,...,...
9988,FUR,FU
9989,TEC,PH
9990,OFF,AP
9991,TEC,AC


In [58]:
print(data_cat_codes['cat_code'].unique())
print(data_cat_codes['sub_cat_code'].unique())

['FUR' 'OFF' 'TEC']
['BO' 'CH' 'LA' 'TA' 'ST' 'FU' 'AR' 'PH' 'BI' 'AP' 'PA' 'AC' 'EN' 'FA'
 'SU' 'MA' 'CO']


In [65]:
catcode_cat = {
    'FUR' : 'Furniture',
    'OFF' :  'Office Supplies',
    'TEC' :  'Technology'
}

subcatcode_subcat = {
    'BO': 'Bookcases',
    'CH':  'Chairs',
    'LA':  'Labels',
    'TA':  'Tables',
    'ST':  'Storage',
    'FU':  'Furnishings',
    'AR':   'Art',
    'PH':   'Phones',
    'BI':   'Binders',
    'AP':  'Appliances',
    'PA':  'Paper',
    'AC':  'Accessories',
    'EN':  'Envelopes',
    'FA':  'Fasteners',
    'SU':  'Supplies',
    'MA':  'Machines',
    'CO':  'Copiers'
    
}



In [66]:
data_cat_codes['Category'] = data['Category']
data_cat_codes['Sub-Category'] = data['Sub-Category']

inconsistent_catcodes_mask = data_cat_codes['Category'] != data_cat_codes['cat_code'].apply(lambda x: catcode_cat[x])
inconsistent_subcatcodes_mask = data_cat_codes['Sub-Category'] != data_cat_codes['sub_cat_code'].apply(lambda x:subcatcode_subcat[x])

print(f"Rows with inconsistent category codes in product id are:\n{data_cat_codes.loc[inconsistent_catcodes_mask]}\n")

print("===============================================================================================================================================")
print(f'\nRows with inconsistent sub-category codes in product id are: \n{data_cat_codes.loc[inconsistent_subcatcodes_mask]}')

Rows with inconsistent category codes in product id are:
Empty DataFrame
Columns: [cat_code, sub_cat_code, Category, Sub-Category]
Index: []


Rows with inconsistent sub-category codes in product id are: 
Empty DataFrame
Columns: [cat_code, sub_cat_code, Category, Sub-Category]
Index: []


Impuation of Missing Values

In [68]:
data.isnull().sum()

Order ID          0
Order Date        0
Ship Date         0
Ship Mode        98
Customer ID       0
Customer Name     0
Segment           0
Country           0
City              0
State             0
Postal Code       0
Region            0
Product ID        0
Category          0
Sub-Category      0
Product Name      0
Sales Price       0
Quantity         17
Discount          0
Profit            0
dtype: int64